### PR TITLE
layout: Make some major improvements to incremental layout to improve CNN.

### PR DESCRIPTION
**Pull request by pcwalton**
CNN is still too slow to be usable, but this is a partial solution.

r? @notriddle (feel free to reassign if you like)

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13346)

### CODE REVIEW

File: components/layout/sequential.rs [hunk 1]
```diff
@@ -78,7 +78,6 @@ pub fn build_display_list_for_subtree(flow_root: &mut Flow,
                                       root_stacking_context: &mut StackingContext,
                                       shared_layout_context: &SharedLayoutContext)
                                       -> Vec<DisplayItem> {
-    flow_root.traverse_preorder(&ComputeAbsolutePositions { layout_context: shared_layout_context });
```
Comment (notriddle) [inline components/layout/sequential.rs]: Just to make sure I'm understanding this right, parallel does not and did not ever perform that traversal?

File: components/layout_thread/lib.rs [hunk 2]
```diff
@@ -928,47 +932,56 @@ impl LayoutThread {
...
-                let root_size = {
-                    let root_flow = flow::base(layout_root);
-                    if rw_data.stylist.viewport_constraints().is_some() {
-                        root_flow.position.size.to_physical(root_flow.writing_mode)
-                    } else {
-                        root_flow.overflow.scroll.size
+                let display_list_needed = query_type.map(reflow_query_type_needs_display_list)
+                                                    .unwrap_or(false);
+                match (data.goal, display_list_needed) {
+                    (ReflowGoal::ForDisplay, _) | (ReflowGoal::ForScriptQuery, true) => {
```
Comment (notriddle) [inline components/layout_thread/lib.rs]: Am I right in guessing that the `query_type` is `Some` if and only if `data.goal == ReflowGoal::ForScriptQuery`?

If so, either add an assertion for that `assert!(query_type.is_some() == data.goal == ReflowGoal::ForScriptQuery)`, or put the query type in the script reflow goal. The latter's more rustic, but might be complicated enough to save for another PR.

**Review by notriddle — COMMENTED** [review]
Review progress:
- [ ] 5a08e8b91031bfaafea666cd103648077b8cdf31
- [X] 28f141b2f5062ce5385a3305ea6a674bddc3434c
- [ ] b393ad3c2722bf37b51869dedd3f389e137cb6c9
- [X] 77fd27623b341e0a9b9943b4bc1257e777e453d5

File: components/layout/sequential.rs (hunk 1 above)
Comment (pcwalton) [inline components/layout/sequential.rs]: Well, it did at one point, but I removed it because it was slower than sequential on all the Web sites I tried. Right now, this pass is always done sequentially.

Comment (notriddle) [inline components/layout/sequential.rs]: Okay, that makes sense.

**Review by notriddle — COMMENTED** [review]
Review progress:
- [X] 5a08e8b
- [X] 28f141b
- [ ] b393ad3
- [X] 77fd276

File: components/layout_thread/lib.rs (hunk 2 above)
Comment (pcwalton) [inline components/layout_thread/lib.rs]: Done.

**Review by notriddle — APPROVED** [review]
Review finished:
- [X] 5a08e8b
- [X] 28f141b
- [X] b393ad3
- [X] 77fd276

As much as I would prefer if this had tests, we don't really have the infrastructure to test performance.

@bors-servo r+ (I wonder if homu will see this?)

**Comment by notriddle** [comment]
Apparently not: @bors-servo r+

**Comment by notriddle** [comment]
servo/homu#61

**Comment by jdm** [comment]
@bors-servo: r-
This still fails tidy.

**Comment by notriddle** [comment]
:facepalm:

**Comment by pcwalton** [comment]
@bors-servo: r=notriddle

The tests were failing because this patch set removed `REFLOW` and `REFLOW_OUT_OF_FLOW` damage bits from replaced absolutely-positioned blocks during block-size assignment, but this was too early as the final positions (and removal of those bits) happen during the `AbsoluteAssignBSizes` traversal instead.

**Comment by jdm** [comment]
@bors-servo: retry
* #13360

**Comment by jdm** [comment]
@bors-servo: retry
#13360

**Comment by jdm** [comment]
Looks like a permanent test failure to me.

**Comment by notriddle** [comment]
An existing bug brought to the surface?

**Comment by pcwalton** [comment]
@bors-servo: r=notriddle

The test was failing because properties that could possibly result in the formation of stacking contexts needed to be marked with `REPOSITION` (since they effectively change stacking-relative positions).

**Comment by notriddle** [comment]
@bors-servo retry

* #12654
